### PR TITLE
kernel: use zstd

### PR DESCRIPTION
**Pull request by Atemu**
###### Motivation for this change

Closes https://github.com/NixOS/nixpkgs/pull/105777

Rebased on top of staging, cleaned up commit messages to fit contributing.md and made a few minor adjustments.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

/marvin opt-in
/status needs_reviewer

### CODE REVIEW

File: nixos/modules/system/boot/stage-1.nix [hunk 1]
```diff
@@ -513,7 +513,13 @@ in
     };
 
     boot.initrd.compressor = mkOption {
-      default = "gzip";
+      default = (
+        if lib.versionAtLeast config.boot.kernelPackages.kernel.version "5.9" then
+          "${lib.getBin pkgs.zstd}/bin/zstd -10"
```
Comment (xaverdh) [inline nixos/modules/system/boot/stage-1.nix:518]: The right way to do this after #97145 is just
```nix
{ 
  default = if lib.versionAtLeast config.boot.kernelPackages.kernel.version "5.9" then "zstd" else "gzip";
}
```

Comment (Atemu) [reply nixos/modules/system/boot/stage-1.nix]: Neat!

File: pkgs/os-specific/linux/kernel/common-config.nix [hunk 2]
```diff
@@ -654,7 +654,10 @@ let
 
       MODULE_COMPRESS    = yes;
       MODULE_COMPRESS_XZ = yes;
-      KERNEL_XZ          = yes;
+
+      # use zstd for kernel compression if newer than 5.9, else xz.
+      KERNEL_XZ          = { optional = true; tristate = whenOlder "5.9" "y";};
+      KERNEL_ZSTD        = { optional = true; tristate = whenAtLeast "5.9" "y";};
```
Comment (lheckemann) [inline pkgs/os-specific/linux/kernel/common-config.nix:660]: I don't think these need to be `optional`? The version conditions should be enough to determine if it's available, and if it's not available in spite of the version theoretically supporting it then there's something funky going on and we should probably error out.

Comment (Atemu) [reply pkgs/os-specific/linux/kernel/common-config.nix]: Yup, works without; much simpler.

**Review by xaverdh — APPROVED** [review]
Looks good to me now. I did not test the `manual-config` changes though.

**Comment by lheckemann** [comment]
@ofborg build linux linux_latest pkgsCross.aarch64-multiplatform.linux pkgsCross.aarch64-multiplatform.linux_latest

**Comment by Atemu** [comment]
Thank you!

**Pull request merged**